File: Sources/WeChatMultiCore/SettingsBackup.swift
import Foundation

/// Export/import of user-tunable settings as JSON, so a new Mac install can
/// start from the previous machine's configuration. Decoupled from
/// `UserDefaults` (works against any `KeyValueStore`) and validated on import.
public enum SettingsBackup {
    public struct Payload: Codable, Equatable {
        public let version: Int
        public let exportedAt: String          // ISO-8601
        public let appVersion: String
        public let slotNames: [String: String] // slot index (string) → name
        public let slotOrder: [Int]
        public let wechatAppPath: String?       // nil = default auto-detection
        public let didShowOnboarding: Bool

        public static let currentVersion = 1

        public init(version: Int, exportedAt: String, appVersion: String,
                    slotNames: [String: String], slotOrder: [Int],
                    wechatAppPath: String?, didShowOnboarding: Bool) {
            self.version = version
            self.exportedAt = exportedAt
            self.appVersion = appVersion
            self.slotNames = slotNames
            self.slotOrder = slotOrder
            self.wechatAppPath = wechatAppPath
            self.didShowOnboarding = didShowOnboarding
        }
    }

    public enum ImportError: LocalizedError, Equatable {
        case malformed(String)
        case unsupportedSchema(found: Int, expected: Int)

        public var errorDescription: String? {
            switch self {
            case .malformed(let detail):
                return "This doesn't look like a valid WeChat Multi settings file. (\(detail))"
            case .unsupportedSchema(let found, let expected):
                return "Unsupported settings file (schema v\(found); this app expects v\(expected))."
            }
        }
    }

    // MARK: - Export

    public static func makePayload(store: KeyValueStore, appVersion: String, now: Date) -> Payload {
        let names = (store.dictionary(forKey: DefaultsKey.slotNames) as? [String: String]) ?? [:]
        let order = (store.array(forKey: DefaultsKey.slotDisplayOrder) as? [Int]) ?? []
        return Payload(
            version: Payload.currentVersion,
            exportedAt: ISO8601DateFormatter().string(from: now),
            appVersion: appVersion,
            slotNames: names,
            slotOrder: order,
            wechatAppPath: store.string(forKey: DefaultsKey.customWeChatPath),
            didShowOnboarding: store.bool(forKey: DefaultsKey.didShowOnboarding)
        )
    }

    public static func exportData(store: KeyValueStore, appVersion: String, now: Date = Date()) throws -> Data {
        let encoder = JSONEncoder()
        encoder.outputFormatting = [.prettyPrinted, .sortedKeys]
        return try encoder.encode(makePayload(store: store, appVersion: appVersion, now: now))
    }

    // MARK: - Import

    /// Decode + validate a settings payload. Throws `ImportError` (not a raw
    /// JSON error) so the UI can present something readable.
    public static func decode(_ data: Data) throws -> Payload {
        let payload: Payload
        do {
            payload = try JSONDecoder().decode(Payload.self, from: data)
        } catch {
            throw ImportError.malformed("couldn't parse JSON")
        }
        guard payload.version == Payload.currentVersion else {
            throw ImportError.unsupportedSchema(found: payload.version,
                                                expected: Payload.currentVersion)
        }
        // Slot-name keys must be positive integers; reject obviously bogus files.
        for key in payload.slotNames.keys {
            guard let n = Int(key), n > 0 else {
                throw ImportError.malformed("invalid slot key “\(key)”")
            }
        }
        return payload
    }

    /// Apply a decoded payload to a store. Sanitizes: a custom WeChat path is
    /// only restored if it currently exists on disk (per `pathExists`), so a
    /// stale path from the old Mac doesn't leave the app pointing at nothing.
    public static func apply(_ payload: Payload,
                             to store: KeyValueStore,
                             pathExists: (String) -> Bool = { FileManager.default.fileExists(atPath: $0) }) {
        store.set(payload.slotNames, forKey: DefaultsKey.slotNames)
        store.set(payload.slotOrder, forKey: DefaultsKey.slotDisplayOrder)
        if let path = payload.wechatAppPath, pathExists(path) {
            store.set(path, forKey: DefaultsKey.customWeChatPath)
        } else {
            store.removeObject(forKey: DefaultsKey.customWeChatPath)
        }
        store.set(payload.didShowOnboarding, forKey: DefaultsKey.didShowOnboarding)
    }

    /// Convenience: decode + apply in one step.
    public static func restore(from data: Data,
                               to store: KeyValueStore,
                               pathExists: (String) -> Bool = { FileManager.default.fileExists(atPath: $0) }) throws {
        let payload = try decode(data)
        apply(payload, to: store, pathExists: pathExists)
    }
}

File: windows/src/WeChatMulti.Core/SettingsBackup.cs
using System.Text.Json;
using System.Text.Json.Serialization;
using WeChatMulti.Core.Settings;

namespace WeChatMulti.Core;

/// <summary>
/// Export/import of user-tunable settings as JSON. Mirrors the macOS
/// SettingsBackup schema (version 1) so the discipline — and the shape — is
/// shared across platforms. Validated on import.
/// </summary>
public static class SettingsBackup
{
    public const int CurrentVersion = 1;

    public sealed record Payload(
        [property: JsonPropertyName("version")] int Version,
        [property: JsonPropertyName("exportedAt")] string ExportedAt,
        [property: JsonPropertyName("appVersion")] string AppVersion,
        [property: JsonPropertyName("instanceNames")] Dictionary<string, string> InstanceNames,
        [property: JsonPropertyName("instanceOrder")] List<int> InstanceOrder,
        [property: JsonPropertyName("wechatAppPath")] string? WeChatAppPath,
        [property: JsonPropertyName("didShowOnboarding")] bool DidShowOnboarding);

    public sealed class ImportException : Exception
    {
        public ImportException(string message) : base(message) { }
    }

    private static readonly JsonSerializerOptions Options = new()
    {
        WriteIndented = true,
        DefaultIgnoreCondition = JsonIgnoreCondition.Never
    };

    // MARK: Export

    public static Payload MakePayload(IKeyValueStore store, string appVersion, DateTimeOffset now) =>
        new(
            Version: CurrentVersion,
            ExportedAt: now.ToString("o"),
            AppVersion: appVersion,
            InstanceNames: new Dictionary<string, string>(
                store.GetStringDict(SettingsKey.InstanceNames) ?? new Dictionary<string, string>()),
            InstanceOrder: (store.GetIntArray(SettingsKey.InstanceOrder) ?? Array.Empty<int>()).ToList(),
            WeChatAppPath: store.GetString(SettingsKey.CustomWeChatPath),
            DidShowOnboarding: store.GetBool(SettingsKey.DidShowOnboarding));

    public static string ExportJson(IKeyValueStore store, string appVersion, DateTimeOffset now) =>
        JsonSerializer.Serialize(MakePayload(store, appVersion, now), Options);

    // MARK: Import

    public static Payload Decode(string json)
    {
        Payload? payload;
        try { payload = JsonSerializer.Deserialize<Payload>(json); }
        catch { throw new ImportException("This doesn't look like a valid WeChat Multi settings file (couldn't parse JSON)."); }

        if (payload is null)
            throw new ImportException("This doesn't look like a valid WeChat Multi settings file (empty).");
        if (payload.Version != CurrentVersion)
            throw new ImportException($"Unsupported settings file (schema v{payload.Version}; this app expects v{CurrentVersion}).");
        foreach (var key in payload.InstanceNames.Keys)
            if (!int.TryParse(key, out var n) || n <= 0)
                throw new ImportException($"Invalid instance key “{key}” in settings file.");
        return payload;
    }

    public static void Apply(Payload payload, IKeyValueStore store, Func<string, bool>? pathExists = null)
    {
        pathExists ??= File.Exists;
        store.SetStringDict(SettingsKey.InstanceNames, payload.InstanceNames);
        store.SetIntArray(SettingsKey.InstanceOrder, payload.InstanceOrder);
        if (payload.WeChatAppPath is { } path && pathExists(path))
            store.SetString(SettingsKey.CustomWeChatPath, path);
        else
            store.Remove(SettingsKey.CustomWeChatPath);
        store.SetBool(SettingsKey.DidShowOnboarding, payload.DidShowOnboarding);
    }

    public static void Restore(string json, IKeyValueStore store, Func<string, bool>? pathExists = null)
        => Apply(Decode(json), store, pathExists);
}
